pkg/migration/generate.go

```diff
@@ -339,13 +382,11 @@ func (h *Handler) constructStorageCluster(ds *appsv1.DaemonSet) *corev1.StorageC
 
 	// Enable CSI
 	csiEnabled := false
-	telemetryEnabled := false
 	for _, c := range ds.Spec.Template.Spec.Containers {
 		switch c.Name {
 		case pxutil.CSIRegistrarContainerName:
 			csiEnabled = true
-		case pxutil.TelemetryContainerName:
-			telemetryEnabled = true
+			cluster.Status.DesiredImages.CSINodeDriverRegistrar = c.Image
```

Comment:
Other CSI images are still missing - 
```
			CSINodeDriverRegistrar:     cluster.Status.DesiredImages.CSINodeDriverRegistrar,
			CSIProvisioner:             cluster.Status.DesiredImages.CSIProvisioner,
			CSIAttacher:                cluster.Status.DesiredImages.CSIAttacher,
			CSIResizer:                 cluster.Status.DesiredImages.CSIResizer,
			CSISnapshotter:             cluster.Status.DesiredImages.CSISnapshotter,
			CSISnapshotController:      cluster.Status.DesiredImages.CSISnapshotController,
			CSIHealthMonitorController: cluster.Status.DesiredImages.CSIHealthMonitorController,
```
I would recommend adding a new method like `addCSISpec` just like the other component methods we have and populating the CSI desired images there, just like how you have done it for telemetry and prometheus.

Reply:
the images anyway will be regenerated based on px version when they upgrade px, why not leave all of them empty and let them be generated based on current px version? 
could you point me to the code/script customer uses to generate the manifest configmap for airgapped env, I want to take a look at it and see if we can reuse that to generate the configmap, i feel there might be glitches in the manifest configmap we generated during migration.

Reply:
Yes, when they upgrade the images will sync, but we in case of air-gapped env we don't want anything to fail. 
If we don't add the other images from the existing CSI deployment, then we will end up using the default as we cannot reach the version manifest. So better to use whatever they are already running instead of using the defaults.

https://docs.portworx.com/portworx-install-with-kubernetes/on-premise/airgapped/

Reply:
which means before upgrading px version, customer needs to run the airgapped script to update the manifest configmap on airgapped env. 

what if we let customer run the airgapped script before migration on airgapped env. they need to do this when doing fresh install or when upgrading, so it does not hurt to run before migration with operator. in this case we dont need to carry those desiredImages in storagecluster -- although this is mostly done now, but still need more comprehensive tests. we just need to output a warning tell them manifest configmap is not present please run airgapped script.

Reply:
Sure, asking them to run the script before migration won't hurt. But, now that this is already handled, we don't need to ask them to use the script anymore, right? 
It is not a lot of change to just read the images from the CSI and prometheus deployments and populate the desired images and configmap anyway.

Reply:
in general if airgapped script is production ready we shall use it, the way we implement here in DesiredImage still have uncertainties and will require more efforts in testing and maintaining going forward. such as

1. migrate while upgrade would not work
2. looks like there are more images to handle, all images in the airgapped list needs to be handled. 

IMAGES="$IMAGES docker.io/portworx/px-enterprise:2.8.1.5"
IMAGES="$IMAGES docker.io/portworx/oci-monitor:2.8.1.5"
IMAGES="$IMAGES docker.io/openstorage/stork:2.7.0"
IMAGES="$IMAGES docker.io/portworx/autopilot:1.3.1"
IMAGES="$IMAGES docker.io/portworx/px-node-wiper:2.5.0"
IMAGES="$IMAGES docker.io/purestorage/ccm-service:3.1.1"
IMAGES="$IMAGES docker.io/portworx/px-lighthouse:2.0.7"
IMAGES="$IMAGES docker.io/portworx/lh-config-sync:2.0.7"
IMAGES="$IMAGES docker.io/portworx/lh-stork-connector:2.0.7"
IMAGES="$IMAGES docker.io/portworx/talisman:1.1.0"
IMAGES="$IMAGES docker.io/portworx/px-operator:1.7.0"
IMAGES="$IMAGES k8s.gcr.io/pause:3.1"
IMAGES="$IMAGES gcr.io/google_containers/kube-controller-manager-amd64:v1.14.10"
IMAGES="$IMAGES gcr.io/google_containers/kube-scheduler-amd64:v1.14.10"
IMAGES="$IMAGES k8s.gcr.io/sig-storage/csi-node-driver-registrar:v2.1.0"
IMAGES="$IMAGES docker.io/openstorage/csi-provisioner:v1.6.1-1"
IMAGES="$IMAGES docker.io/openstorage/csi-attacher:v1.2.1-1"
IMAGES="$IMAGES k8s.gcr.io/sig-storage/csi-resizer:v1.1.0"
IMAGES="$IMAGES docker.io/openstorage/csi-snapshotter:v1.2.2-1"
IMAGES="$IMAGES k8s.gcr.io/sig-storage/snapshot-controller:v3.0.3"
IMAGES="$IMAGES quay.io/prometheus/prometheus:v2.7.1"
IMAGES="$IMAGES quay.io/coreos/prometheus-operator:v0.34.0"
IMAGES="$IMAGES quay.io/coreos/configmap-reload:v0.0.1"
IMAGES="$IMAGES quay.io/coreos/prometheus-config-reloader:v0.34.0"
IMAGES="$IMAGES quay.io/prometheus/alertmanager:v0.17.0"

Reply:
- I went through the list already. The only images that are not accounted for are the pvc controller and stork scheduler images. Rest all the images will be covered by the desired images.
- We are not going to recommend upgrade during the migration at least for now. 

Basically I am saying there is no harm in handling the rest of the images if we can. It will definitely make the experience better. We can definitely recommend running the air-gapped script before the migration, but even without it we can handle the migration. Additionally we also, take care of creating the px-version configmap for them.

Reply:
with the issues above and potentially overlooked issues, I would prefer just stop migration and output a warning let user run airgapped script first. we dont want to run into various issues with this implementation and maintaining it. 

1. how about pause image, k8s controller images etc
2. it's not blocked, chances are still there

what do you think?

Reply:
Pause image has never changed. K8s controller image == pvc controller image. Version for this image is taken from the k8s version. But it is possible that customers updated the k8s version but did not change the k8s controller and k8s scheduler images (stork-scheduler) in their deployments.

I agree that we should throw a warning and not proceed with the migration. From what comes to my mind, we should stop the migration and ask users to run the air-gapped script when - 
- stork-scheduler image version != current k8s version
- pvc controller image version != current k8s version 
- all images don't have the same custom image registry prefix 

If none of the above issues are encountered we should be able to proceed with the migration.